utils/s3/client.cc

```diff
@@ -592,6 +592,110 @@ class client::multipart_upload {
     unsigned parts_count() const noexcept { return _part_etags.size(); }
 };
 
+class client::copy_s3_object final : multipart_upload {
+public:
+    copy_s3_object(shared_ptr<client> cln, sstring source_object, sstring target_object, size_t part_size, std::optional<tag> tag, abort_source* as)
+        : multipart_upload(std::move(cln), std::move(target_object), std::move(tag), as)
+        , _max_copy_part_size(part_size)
+        , _source_object(std::move(source_object)) {
+        assert(_max_copy_part_size > 0 && _max_copy_part_size <= _default_copy_part_size);
+    }
+
+    copy_s3_object(shared_ptr<client> cln, sstring source_object, sstring target_object, std::optional<tag> tag, abort_source* as)
+        : copy_s3_object(std::move(cln), std::move(source_object), std::move(target_object), _default_copy_part_size, std::move(tag), as) {}
+
+    future<> copy() {
+        auto source_size = co_await _client->get_object_size(_source_object);
+        if (source_size <= _max_copy_part_size) {
+            co_await copy_put();
+        } else {
+            co_await copy_multipart(source_size);
+        }
+    }
+
+private:
+    future<> copy_put() {
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        if (_tag) {
+            req._headers["x-amz-tagging"] = seastar::format("{}={}", _tag->key, _tag->value);
+        }
+        req._headers["x-amz-copy-source"] = _source_object;
+
+        co_await _client->make_request(std::move(req), ignore_reply, http::reply::status_type::ok, _as);
+    }
+
+    future<> copy_multipart(size_t source_size) {
+        co_await start_upload();
+        auto part_size = _max_copy_part_size;
+        std::exception_ptr ex;
+
+        try {
+            for (size_t offset = 0; offset < source_size; offset += part_size) {
+                part_size = std::min(source_size - offset, part_size);
+                co_await copy_part(offset, part_size);
+            }
+            // Here we are going to finalize the upload and close the _bg_flushes, in case an exception is thrown the
+            // gate will be closed and the upload will be aborted. See below.
+            co_await finalize_upload();
+        } catch (...) {
+            ex = std::current_exception();
+        }
+
+        if (ex) {
+            if (!_bg_flushes.is_closed()) {
+                co_await _bg_flushes.close();
+            }
+            co_await abort_upload();
+            std::rethrow_exception(ex);
+        }
+    }
+
+    future<> copy_part(size_t offset, size_t part_size) {
```

Comment:
This doesn't seem to differ much from `client::multipart_upload::upload_part(std::unique_ptr<upload_sink> piece_ptr)`. With the exception that you don't have the piece_ptr thing at hand. It's worth generalizing this thing